src/ImageSharp/Formats/Webp/Lossless/LosslessUtils.cs

```diff
@@ -759,28 +760,184 @@ public static void BundleColorMap(Span<byte> row, int width, int xBits, Span<uin
         /// <returns>Shanon entropy.</returns>
         public static float CombinedShannonEntropy(Span<int> x, Span<int> y)
         {
-            double retVal = 0.0d;
-            uint sumX = 0, sumXY = 0;
-            for (int i = 0; i < 256; i++)
+#if SUPPORTS_RUNTIME_INTRINSICS
+            if (Avx2.IsSupported)
             {
-                uint xi = (uint)x[i];
-                if (xi != 0)
+                double retVal = 0.0d;
+                Span<int> tmp = stackalloc int[8];
+                ref int xRef = ref MemoryMarshal.GetReference(x);
+                ref int yRef = ref MemoryMarshal.GetReference(y);
+                Vector256<int> sumXY256 = Vector256<int>.Zero;
+                Vector256<int> sumX256 = Vector256<int>.Zero;
+                ref int tmpRef = ref MemoryMarshal.GetReference(tmp);
+                for (nint i = 0; i < 256; i += 8)
                 {
-                    uint xy = xi + (uint)y[i];
-                    sumX += xi;
-                    retVal -= FastSLog2(xi);
-                    sumXY += xy;
-                    retVal -= FastSLog2(xy);
+                    Vector256<int> xVec = Unsafe.As<int, Vector256<int>>(ref Unsafe.Add(ref xRef, i));
+                    Vector256<int> yVec = Unsafe.As<int, Vector256<int>>(ref Unsafe.Add(ref yRef, i));
+
+                    // Check if any X is non-zero: this actually provides a speedup as X is usually sparse.
+                    int mask = Avx2.MoveMask(Avx2.CompareEqual(xVec, Vector256<int>.Zero).AsByte());
+                    if (mask != -1)
+                    {
+                        Vector256<int> xy256 = Avx2.Add(xVec, yVec);
+                        sumXY256 = Avx2.Add(sumXY256, xy256);
+                        sumX256 = Avx2.Add(sumX256, xVec);
+
+                        // Analyze the different X + Y.
+                        Unsafe.As<int, Vector256<int>>(ref tmpRef) = xy256;
+                        if (tmpRef != 0)
+                        {
+                            retVal -= FastSLog2((uint)tmpRef);
+                            if (Unsafe.Add(ref xRef, i) != 0)
+                            {
+                                retVal -= FastSLog2((uint)Unsafe.Add(ref xRef, i));
+                            }
+                        }
+
+                        if (Unsafe.Add(ref tmpRef, 1) != 0)
```

Comment:
Note: I have tried to put those repeating if statements into own method calls, but profiling has shown that this makes it actually slower even with Aggressive Inlining.

Reply:
Couldn't it be a loop? It looks incremental 0-7 to me. 

However, my money says there's something clever that can be done with masking here to determine if each element != 0 and apply the diff as a single operation. I'm sure I've seen similar before.

Reply:
Simd? You can check each element in a vector without if checks at all for `tmpRef` and `xRef` vectors.

You can also remove if checks with log2  precalculation for each case and simply multiplying that log2 vectors with comparison mask values. This may or may not be faster than if-checks - depends ong FastSLog2 implementation.

Reply:
The original SSE2 version of this uses macros. I tried to keep it as similar as possible.
Because of that, I thought a loop might be also slower, but I will give it a try and test it.

`FastSLog2` is in most cases (>90% just a lookup table), but still most of the time is spend here.

Reply:
I guess loop can further degrade branch predictor history. Masking is still possible but that would require some serious code rewrite for AVX branch.

Reply:
I have done profile tests now with loops several times and it makes it noticeable slower.

I am not really sure that it is worth trying to optimize it further, if it would take a huge amount of work to do so. As can be seen here:
![shannon_avx2](https://user-images.githubusercontent.com/38701097/143114215-4e0123da-f0ad-4a60-a554-8fa8b1e93e53.png)

The main time consumer is `FastSLog2`.

Reply:
I honestly have no idea how webp works but as far as I understand checked value is actually pretty random, adding for-loop on top of it may screw up branch predictor with yet another stable (always true for 8 iterations) if-check - that's what's most likely causing performance drop.

All of extra if-checks can be removed via simd masks but it's a question whether it'd be faster than current implementation.

Reply:
I had a previous SSE2 version which used masks, if you look at commit: ed8bd615f2be3cafd1a23782e9f7d07c6375d967, but this was not better then the current approach. Maybe, if we could create a AVX version of that, it could be better, but I am also not sure if it really can beat the current one.

Reply:
It's actually easier to describe it in code than in human language. I'll probably try to implement it after this gets merged.

Reply:
@br3aker thanks, but my advice would be, to only do it, if you think it would be easy for you and not to much work. As I said before, I am really unsure how much we can gain from this.